internal/server/v2/observation/calculation_util_test.go

```diff
@@ -152,3 +154,194 @@ func TestFindObservationResponseHoles(t *testing.T) {
 		}
 	}
 }
+
+func TestComputeLeafSeries(t *testing.T) {
+	for _, c := range []struct {
+		inputResp *pbv2.ObservationResponse
+		formula   *VariableFormula
+		want      *VariableFormula
+	}{{
+		&pbv2.ObservationResponse{
+			ByVariable: map[string]*pbv2.VariableObservation{
+				"Count_Person": {ByEntity: map[string]*pbv2.EntityObservation{
+					"geoId/01": {OrderedFacets: []*pbv2.FacetObservation{
+						{
+							FacetId: "1",
+							Observations: []*pb.PointStat{{
+								Date:  "1",
+								Value: proto.Float64(1),
+							}},
+						},
+						{
+							FacetId: "2",
+							Observations: []*pb.PointStat{{
+								Date:  "2",
+								Value: proto.Float64(2),
+							}},
+						},
+					}},
+				}},
+			},
+			Facets: map[string]*pb.Facet{
+				"1": {
+					ObservationPeriod: "P1M",
+				},
+				"2": {
+					MeasurementMethod: "US_Census",
+					ObservationPeriod: "P1Y",
+				},
+			},
+		},
+		&VariableFormula{
+			LeafData: map[string]*ASTNode{
+				"Count_Person": {StatVar: "Count_Person"},
+				"Count_Person[mm=US_Census;p=P1Y]": {
+					StatVar: "Count_Person",
+					Facet: &pb.Facet{
+						MeasurementMethod: "US_Census",
+						ObservationPeriod: "P1Y",
+					},
+				},
+			},
+		},
+		&VariableFormula{
+			LeafData: map[string]*ASTNode{
+				"Count_Person": {
+					StatVar: "Count_Person",
+					CandidateSeries: map[string]map[string][]*pb.PointStat{
+						"geoId/01": {
+							"1": {{
+								Date:  "1",
+								Value: proto.Float64(1),
+							}},
+							"2": {{
+								Date:  "2",
+								Value: proto.Float64(2),
+							}},
+						},
+					},
+				},
+				"Count_Person[mm=US_Census;p=P1Y]": {
+					StatVar: "Count_Person",
+					Facet: &pb.Facet{
+						MeasurementMethod: "US_Census",
+						ObservationPeriod: "P1Y",
+					},
+					CandidateSeries: map[string]map[string][]*pb.PointStat{
+						"geoId/01": {
+							"2": {{
+								Date:  "2",
+								Value: proto.Float64(2),
+							}},
+						},
+					},
+				},
+			},
+		},
+	}} {
+		computeLeafSeries(c.inputResp, c.formula)
+		if ok := reflect.DeepEqual(c.formula, c.want); !ok {
+			t.Errorf("computeLeafSeries = %v, want %v",
+				c.formula, c.want)
+		}
+	}
+}
+
+func TestEvalBinaryExpr(t *testing.T) {
+	for _, c := range []struct {
+		x    map[string]map[string][]*pb.PointStat
+		y    map[string]map[string][]*pb.PointStat
+		op   token.Token
+		want map[string]map[string][]*pb.PointStat
+	}{
+		{
+			map[string]map[string][]*pb.PointStat{
+				"geoId/01": {"facetId2": {
```

Comment:
for all recent changes, i wonder if you want to have testdata/xx.proto file to contain the test data and load them to use here. Might be easier to write parameterized tests and maintain them..

up to you and could be a future change.

Reply:
Good idea! I parameterized the TestEvalBinaryExpr somewhat and can split this up further as a follow up